osu.Server.Spectator/Hubs/SpectatorHub.cs

```diff
@@ -18,7 +20,11 @@ public SpectatorHub([NotNull] IDistributedCache cache)
 
         public async Task BeginPlaySession(SpectatorState state)
         {
-            await UpdateLocalUserState(state);
+            using (var usage = await GetOrCreateLocalUserState())
+            {
+                usage.Item ??= new SpectatorClientState(Context.ConnectionId, CurrentContextUserId);
+                usage.Item.State = state;
+            }
```

Comment:
Same as the other comment - broadcasting `state` to all users without a defensive copy inside this block is also similarly potentially unsafe (as someone has potentially grabbed the semaphore on the state and potentially started making changes to it - but this one is way smaller impact as there are no direct mutations to the state under this reference as far as I can see, the references are being replaced entirely on the entity if there's a change).

Reply:
Yep, but that will eventually be handled in a better way.